### PR TITLE
Multicategory axis type

### PR DESCRIPTION
**Pull request by etpinard**
(attempts to) resolves https://github.com/plotly/plotly.js/issues/2799 by adding a :sparkles: new :sparkles: axis type: `'multicategory'` - which displays two (maybe more later) axis labels levels along with "dividers" in-between each category group. 

**N.B.** This branch is up-to-date with the latest master that includes `uirevision`.

--- 

Things to TODO:

- [x] add support for `type: 'multicategory'` on y-axes
- [x] add support to heatmap and contours
- [x] add support in `histogram*` traces
- <strike>[ ] add some way to configure ticks/dividers per category level</strike> - **leave for later**
- [x] improve `axis.automargin` test coverage, **partly done in:** https://github.com/plotly/plotly.js/pull/3300/commits/80f47014b8bbfb0743fbd81ebd56295d629e3be0, more to come in https://github.com/plotly/plotly.js/issues/3305
- [x] test as many edge cases I can think of

Things skipped:

- `ticktext` / `tickvals` are not coerced on `type: 'multicategory'` axes
- support for `transforms`

cc @plotly/plotly_js @nicolaskruchten

### CODE REVIEW

File: test/image/mocks/multicategory.json [hunk 1]
```diff
@@ -0,0 +1,28 @@
+{
+  "data": [
+    {
+      "type": "bar",
+      "x": [
+        ["2017", "2017", "2017", "2017", "2018", "2018", "2018"],
+        ["q1", "q2", "q3", "q4", "q1", "q2", "q3" ]
```
Comment (etpinard) [inline test/image/mocks/multicategory.json]: Category "group" in `x[0]`, then the inner in `x[1]`.

File: src/plots/cartesian/axis_autotype.js [hunk 2]
```diff
@@ -81,3 +82,10 @@ function category(a) {
 
     return curvecats > curvenums * 2;
 }
+
+function multiCategory(a) {
+    return (
+        Array.isArray(a[0]) && Array.isArray(a[1]) &&
+        (category(a[0]) || category(a[1]))
+    );
```
Comment (etpinard) [inline src/plots/cartesian/axis_autotype.js]: This is we detect multicategory trace used to auto-type the axes. I think someone thought of a better way during yesterday's meeting. I forgot what that was though.

File: src/traces/bar/cross_trace_calc.js [hunk 3]
```diff
@@ -126,6 +126,8 @@ function initBase(gd, pa, sa, calcTraces) {
         // time. But included here for completeness.
         var scalendar = trace.orientation === 'h' ? trace.xcalendar : trace.ycalendar;
 
+        // TODO skip for multicategory size axis?
```
Comment (etpinard) [inline src/traces/bar/cross_trace_calc.js]: Related to https://github.com/plotly/plotly.js/issues/2391#issuecomment-444545411

File: src/plots/cartesian/axes.js [hunk 4]
```diff
@@ -1163,19 +1165,21 @@ function formatCategory(ax, out) {
...
+    var tt = cats[1] === undefined ? '' : String(cats[1]);
+    var tt2 = cats[0] === undefined ? '' : String(cats[0]);
 
-        out.xbnd = [
-            inbounds(out.x - 0.5),
-            inbounds(out.x + ax.dtick - 0.5)
-        ];
+    if(hover) {
+        // TODO is this what we want?
+        out.text = tt2 + ' - ' + tt;
```
Comment (etpinard) [inline src/plots/cartesian/axes.js]: This looks like:

peek 2018-12-05 19-02


I could use a second opinion here.

File: src/plots/cartesian/axis_defaults.js [hunk 5]
```diff
@@ -91,9 +91,19 @@ module.exports = function handleAxisDefaults(containerIn, containerOut, coerce,
     if(options.automargin) coerce('automargin');
 
     if(!options.noTickson &&
-        containerOut.type === 'category' && (containerOut.ticks || containerOut.showgrid)) {
+        (containerOut.type === 'category' || containerOut.type === 'multicategory') &&
+        (containerOut.ticks || containerOut.showgrid)
+    ) {
         coerce('tickson');
```
Comment (etpinard) [inline src/plots/cartesian/axis_defaults.js]: Should `tickson: 'boundaries'` by the default on `type: 'multicategory` axes?

File: src/plots/cartesian/layout_attributes.js [hunk 6]
```diff
@@ -666,6 +666,40 @@ module.exports = {
         editType: 'ticks',
         description: 'Sets the width (in px) of the zero line.'
     },
+
+    showdividers: {
```
Comment (etpinard) [inline src/plots/cartesian/layout_attributes.js]: Adding axis-wide `showdividers`, `dividercolor` and `dividerwidth`. Would this be enough for the first iteration, or per-category-level setting required?

File: src/plots/cartesian/axis_autotype.js (hunk 2 above)
Comment (etpinard) [reply src/plots/cartesian/axis_autotype.js]: :zap: that second line.

File: src/plots/cartesian/axis_defaults.js (hunk 5 above)
Comment (etpinard) [reply src/plots/cartesian/axis_defaults.js]: yes!

File: src/plots/cartesian/axis_autotype.js (hunk 2 above)
Comment (etpinard) [reply src/plots/cartesian/axis_autotype.js]: **Edit:** Removing that `&& (category(a[0]) || category(a[1])` breaks some `carpet` plots which allow 2D `x` and `y` arrays.

Comment (etpinard) [reply src/plots/cartesian/axis_autotype.js]: .. so should we make a special case for `carpet` traces? Or require that one of the two inner arrays have categories.

File: src/plots/cartesian/axes.js (hunk 4 above)
Comment (etpinard) [reply src/plots/cartesian/axes.js]: in-person @alexcjohnson mentioned that a simple space ` ` could perhaps be better than the current ` - ` . Thoughts?

File: src/plots/cartesian/axis_autotype.js (hunk 2 above)
Comment (alexcjohnson) [reply src/plots/cartesian/axis_autotype.js]: Carpet is definitely special, nobody would ever really mistake carpets for real multicategory. But for extensibility perhaps a trace module category “noMulticategory” or something.

File: src/plots/cartesian/axis_defaults.js (hunk 5 above)
Comment (etpinard) [reply src/plots/cartesian/axis_defaults.js]: done in https://github.com/plotly/plotly.js/pull/3300/commits/2bb1388b336058e5892b68a9a5e58b453bbb3ae9

File: src/traces/bar/cross_trace_calc.js (hunk 3 above)
Comment (etpinard) [reply src/traces/bar/cross_trace_calc.js]: skipped in https://github.com/plotly/plotly.js/pull/3300/commits/3eac3384e7a40a3873d74957ecb105f1ae60a156

File: src/plots/cartesian/axis_autotype.js (hunk 2 above)
Comment (etpinard) [reply src/plots/cartesian/axis_autotype.js]: Done in https://github.com/plotly/plotly.js/pull/3300/commits/3746e03dc9e970062a307c0f348d57573b056d63

**Comment by etpinard** [comment]
@alexcjohnson 

I'll take a deep dive into why `Drawing.bBox` doesn't work well with rotated text labels, and I'd like to fix https://github.com/plotly/plotly.js/issues/2796 (which seems to have an effect on dividers too), but I think it's time to tag this thing as `reviewable`.

File: src/plots/cartesian/axis_autotype.js [hunk 7]
```diff
@@ -81,3 +84,10 @@ function category(a) {
 
     return curvecats > curvenums * 2;
 }
+
+// very-loose requirements for multicategory,
+// trace modules that should never auto-type to multicategory
+// should be declared with 'noMultiCategory'
+function multiCategory(a) {
+    return Array.isArray(a[0]) && Array.isArray(a[1]);
```
Comment (alexcjohnson) [inline src/plots/cartesian/axis_autotype.js]: We decided these should be `isArrayOrTypedArray`, didn't we?

File: src/traces/bar/cross_trace_calc.js [hunk 8]
```diff
@@ -126,9 +126,14 @@ function initBase(gd, pa, sa, calcTraces) {
         // time. But included here for completeness.
         var scalendar = trace.orientation === 'h' ? trace.xcalendar : trace.ycalendar;
 
+        // 'base' on categorical axes makes no sense
+        var d2c = sa.type === 'category' || sa.type === 'multicategory' ?
+            function() { return null; } :
+            sa.d2c;
```
Comment (alexcjohnson) [inline src/traces/bar/cross_trace_calc.js]: Ha clever - not the most performant solution, but possibly the least code, and since this is a failsafe rather than a real use case that's the more important criterion!

File: src/traces/box/calc.js [hunk 9]
```diff
@@ -178,7 +178,10 @@ function getPos(trace, posLetter, posAxis, val, num) {
         pos0 = num;
     }
 
-    var pos0c = posAxis.d2c(pos0, 0, trace[posLetter + 'calendar']);
+    var pos0c = posAxis.type === 'multicategory' ?
+        posAxis.r2c(pos0) :
```
Comment (alexcjohnson) [inline src/traces/box/calc.js]: Does it work to specify `x0: ['2016', 'day 1']` or does `x0` need to be a number? I'm not really sure the latter is something we want to encourage, and the former seems to me fairly important to support.

I see your mock with the violin halfway between the two inner categories, which of course isn't possible using the categories themselves, I feel like if someone really wants that we should add something like `boxpos`, ie an analog of `pointpos` to allow you to shift where the box is drawn, but still have it explicitly anchored to a specific category rather than to our serialized category numbers.

File: src/traces/histogram/calc.js [hunk 10]
```diff
@@ -327,7 +328,7 @@ function calcAllAutoBins(gd, trace, pa, mainData, _overlayEdgeCase) {
             Lib.aggNums(Math.min, null, pos0);
 
         var dummyAx = {
-            type: pa.type === 'category' ? 'linear' : pa.type,
+            type: (pa.type === 'category' || pa.type === 'multicategory') ? 'linear' : pa.type,
```
Comment (alexcjohnson) [inline src/traces/histogram/calc.js]: this is showing up a lot - I almost want to make something like `ax._isAnyCategory` (or perhaps `ax._isOrdinal`?) during `supplyDefaults`. Non-blocking, just something to consider.

File: src/traces/histogram2d/sample_defaults.js [hunk 11]
```diff
@@ -6,24 +6,26 @@
...
     var x = coerce('x');
     var y = coerce('y');
+    var xlen = Lib.minRowLength(x);
+    var ylen = Lib.minRowLength(y);
 
     // we could try to accept x0 and dx, etc...
     // but that's a pretty weird use case.
     // for now require both x and y explicitly specified.
-    if(!(x && x.length && y && y.length)) {
+    if(!(x && xlen && y && ylen)) {
```
Comment (alexcjohnson) [inline src/traces/histogram2d/sample_defaults.js]: you already did this elsewhere... but one of the nice things about your `Lib.minRowLength` is you can just say `if(!(xlen && ylen))` right?

File: src/plots/cartesian/axes.js [hunk 12]
```diff
@@ -2370,40 +2495,75 @@ axes.drawLabels = function(gd, ax, opts) {
...
+ *  - {string} showdividers
+ *  - {number} dividerwidth
+ *  - {string} dividercolor
+ * @param {object} opts
+ * - {array of object} vals (calcTicks output-like)
+ * - {d3 selection} layer
+ * - {fn} path
+ * - {fn} transFn
+ */
+function drawDividers(gd, ax, opts) {
```
Comment (alexcjohnson) [inline src/plots/cartesian/axes.js]: Can we remove ticks that would overlap dividers? I guess if we had inside ticks they should still be there, but not outside ticks.

File: test/image/mocks/multicategory_histograms.json [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+        [ "a", "a", "a", "a", "a", "a", "b", "b", "c", "c", "c", "c" ]
+      ],
+      "yaxis": "y2",
+      "marker": {
+        "color": "#008080"
+      }
+    },
+    {
+      "type": "histogram",
+      "name": "hist-y",
```
Comment (alexcjohnson) [inline test/image/mocks/multicategory_histograms.json]: This kind of plot would really benefit from `dividerlines` 😏

File: test/jasmine/tests/hover_label_test.js [hunk 14]
```diff
@@ -2055,6 +2055,67 @@ describe('hover on fill', function() {
...
+            x: [['2018', '2018', '2019', '2019'], ['a', 'b', 'a', 'b']],
+            y: [1, 2, -1, 3]
+        }], {
+            bargap: 0,
+            width: 400,
+            height: 400
+        })
+        .then(function() { _hover(200, 200); })
+        .then(function() {
+            assertHoverLabelContent({ nums: '−1', axis: '2019 - a' });
```
Comment (alexcjohnson) [inline test/jasmine/tests/hover_label_test.js]: What do we get in event data for the multicategory value? `x: ['2019', 'a']` or something else?

File: src/plots/cartesian/axes.js (hunk 12 above)
Comment (alexcjohnson) [reply src/plots/cartesian/axes.js]: hmm this would get complicated with mirrored ticks... dividers won't mirror, right? So ideally the mirrored ones would still appear, but since we draw all ticks for a given x value as a single path containing all the mirrors, hiding just the main one would require a different path when we need to exclude the main one.

File: src/plots/cartesian/axis_autotype.js (hunk 7 above)
Comment (etpinard) [reply src/plots/cartesian/axis_autotype.js]: Good eye -> https://github.com/plotly/plotly.js/pull/3300/commits/a9eca0951f568d4202dcb0fff8f5546948c27889

File: src/plots/cartesian/set_convert.js [hunk 15]
```diff
@@ -255,6 +300,29 @@ module.exports = function setConvert(ax, fullLayout) {
             return ensureNumber(v);
         };
     }
+    else if(ax.type === 'multicategory') {
+        // ax.d2c = ax.d2l = setMultiCategoryIndex;
```
[the remark this answers is not in the document]
Comment (etpinard) [reply src/plots/cartesian/set_convert.js]: Explanation added in https://github.com/plotly/plotly.js/pull/3300/commits/3bc03e5688a7c45c550255ad90e4c084fd0f9c20

File: src/traces/bar/cross_trace_calc.js (hunk 8 above)
Comment (etpinard) [reply src/traces/bar/cross_trace_calc.js]: > that's the more important criterion!

I'm glad you agree!

File: src/traces/box/calc.js (hunk 9 above)
Comment (etpinard) [reply src/traces/box/calc.js]: > Does it work to specify `x0: ['2016', 'day 1']`

This works now in https://github.com/plotly/plotly.js/pull/3300/commits/5c8598121f6a04b6193e5aab65de6916701eb6af

Should we disallow number `x0` that falls in-between categories?

File: src/traces/histogram2d/sample_defaults.js (hunk 11 above)
Comment (etpinard) [reply src/traces/histogram2d/sample_defaults.js]: done in https://github.com/plotly/plotly.js/pull/3300/commits/627f24b68052e1cda45fea65fcf20901309bf303

File: test/image/mocks/heatmap_multicategory.json [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+    "z": [
+      [ 0.1304, 1.046, 2.45, 3.226, 4.34, 5.40, 6.05 ],
+      [ 0.3515, 1.326, 2.18, 3.26, 4.41, 5.25, 6.11 ],
+      [ 0.3994, 1.167, 2.09, 3.306, 4.305, 5.35, 6.00 ],
+      [ 0.32, 1.44, 2.303, 3.115, 4.49, 5.25, 6.46 ],
+      [ 0.297, 1.295, 2.49, 3.428, 4.13, 5.41, 6.38 ],
+      [ 0.0197, 1.274, 2.407, 3.22, 4.47, 5.44, 6.28 ],
+      [ 0.4446, 1.223, 2.367, 3.253, 4.385, 5.08, 6.19 ],
+      [ 0.4602, 1.2256, 2.3356, 3.0667, 4.498, 5.411, 6.29 ],
+      [ 0.304, 1.465, 2.474, 3.05, 4.38, 5.245, 6.12 ]
```
[the remark this answers is not in the document]
Comment (etpinard) [reply test/image/mocks/heatmap_multicategory.json]: Done in https://github.com/plotly/plotly.js/pull/3300/commits/a25c5e3cf23bd5af3f2bf45e3f161d7569fdf9fd

File: test/jasmine/tests/hover_label_test.js (hunk 14 above)
Comment (etpinard) [reply test/jasmine/tests/hover_label_test.js]: > x: ['2019', 'a']

We get that! :lock: down in https://github.com/plotly/plotly.js/pull/3300/commits/e71b7df03a2ef5755561eea86a9accc6beede137

File: src/plots/cartesian/axes.js (hunk 12 above)
Comment (etpinard) [reply src/plots/cartesian/axes.js]: > Can we remove ticks that would overlap dividers?

Yeah, we could. I tried to "solve" the problem by placing dividers below all ticks, but removing outside / `tickon:'boundaries'` ticks would be better. I'll see what I can do.



> dividers won't mirror, right?

Correct, dividers don't mirror.

Comment (etpinard) [reply src/plots/cartesian/axes.js]: See my attempt in https://github.com/plotly/plotly.js/compare/multicategory...multicategory-polishing

File: src/traces/box/calc.js (hunk 9 above)
Comment (alexcjohnson) [reply src/traces/box/calc.js]: > Should we disallow number `x0` that falls in-between categories?

Personally I'd be in favor of not allowing `x0` to be a serial number at all. It can be a number on a regular category axis, but that number is treated as a category. So allowing a serial number would be unique to `multicategory` and I suspect more likely confusing than useful.

File: src/plots/cartesian/axes.js (hunk 4 above)
Comment (alexcjohnson) [reply src/plots/cartesian/axes.js]: I still think a single space is my vote, but it's not a big deal to me either way. @nicolaskruchten @chriddyp do you have a preference?

File: src/traces/box/calc.js (hunk 9 above)
Comment (etpinard) [reply src/traces/box/calc.js]: Are you ok with -> https://github.com/plotly/plotly.js/commit/27d74d8060a2a026ee08c531d24672d2df4af8c1

File: src/plots/cartesian/axes.js (hunk 4 above)
Comment (etpinard) [reply src/plots/cartesian/axes.js]: ... and note that `hovertemplate` _just work_<sup>TM</sup> with multicategory axes:

https://github.com/plotly/plotly.js/blob/a25c5e3cf23bd5af3f2bf45e3f161d7569fdf9fd/test/jasmine/tests/hover_label_test.js#L2096

so one can easily customize the appearance.

Comment (nicolaskruchten) [reply src/plots/cartesian/axes.js]: nice re hovertemplates! I favour the hyphen personally

File: src/plots/cartesian/axes.js (hunk 12 above)
Comment (etpinard) [reply src/plots/cartesian/axes.js]: In https://github.com/plotly/plotly.js/pull/3300/commits/5da38513fc5224eaa29b7bf101d921732fe64c8e

Comment (etpinard) [reply src/plots/cartesian/axes.js]: with new test mock in https://github.com/plotly/plotly.js/pull/3300/commits/204b2e803bbc93bbaf2a909025d24df0b6f67fa9

File: src/traces/box/calc.js (hunk 9 above)
Comment (etpinard) [reply src/traces/box/calc.js]: done in https://github.com/plotly/plotly.js/pull/3300/commits/4b82426021d1e0957f45800a65a36328a5d6ba4a

**Comment by alexcjohnson** [comment]
💃 love it.

**Comment by Braintelligence** [comment]
Just to make 100% sure:
Is it somehow possible to combine this with `xaxis.type = "date"` or `linear` if you want your datapoint to be between two ticks (divider style) and still keep the advantages of these axis types? (scalability and so on)
As it stands now it seems that `showdividers` is not compatible with `xaxis.type = "date"` to me.

**Comment by etpinard** [comment]
> As it stands now it seems that `showdividers` is not compatible with `xaxis.type = "date"` to me.

That's correct. Dividers are only available for `type: 'multicategory'` axes.

**Comment by LJaschinski** [comment]
Hello, do you know when 'multicategory' type will displays more than two axis ? Is there a work in progress about it ?

**Comment by elmadj** [comment]
Hi is there support for something similar to bargap and bargroupgap with multicategory axes? The goal would be to have a different gaps between bars of an outer category and bars of an inner category. Thanks !
